src/array.jl

```diff
@@ -141,18 +143,15 @@ CategoricalArray{T}(m::Int, n::Int; ordered=false) where {T} =
 CategoricalArray{T}(m::Int, n::Int, o::Int; ordered=false) where {T} =
     CategoricalArray{T}((m, n, o), ordered=ordered)
 
-CategoricalArray{CategoricalValue{T, R}, N, R}(dims::NTuple{N,Int};
-                                               ordered=false) where {T, N, R} =
-    CategoricalArray{T, N, R}(dims, ordered=ordered)
-CategoricalArray{CategoricalValue{T}, N, R}(dims::NTuple{N,Int};
-                                            ordered=false) where {T, N, R} =
-    CategoricalArray{T, N, R}(dims, ordered=ordered)
-CategoricalArray{CategoricalValue{T, R}, N}(dims::NTuple{N,Int};
-                                            ordered=false) where {T, N, R} =
-    CategoricalArray{T, N, R}(dims, ordered=ordered)
-CategoricalArray{CategoricalValue{T}, N}(dims::NTuple{N,Int};
-                                         ordered=false) where {T, N} =
-    CategoricalArray{T, N}(dims, ordered=ordered)
+function CategoricalArray{T, N, R}(dims::NTuple{N,Int};
+                                   ordered=false) where {T<:CatValue, N, R<:Integer}
+    V = unwrap_catvalue_type(T)
+    CategoricalArray{V, N, R}(zeros(R, dims), CategoricalPool{V, R, T}(ordered))
```

Comment:
I think you can call `CategoricalArray{T, N, R}(dims, ordered=ordered)`. That way all constructors fall back to a single one, that makes it easier to understand how all this mess is organized.